### PR TITLE
expression: impl cast as array function

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40128

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -467,9 +468,79 @@ func (b *castJSONAsArrayFunctionSig) evalJSON(row chunk.Row) (res types.BinaryJS
...
+		if err != nil {
+			return types.BinaryJSON{}, false, err
+		}
+		arrayVals = append(arrayVals, item)
+	}
+	return types.CreateBinaryJSON(arrayVals), false, nil
+}
 
-	return types.BinaryJSON{}, false, nil
+func convertJSON2Tp(tp *types.FieldType) func(*stmtctx.StatementContext, types.BinaryJSON) (any, error) {
```
Comment (xhebox) [inline expression/builtin_cast.go:498]: Can not it be just a switch of `tp.EvalType()`?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -467,9 +468,79 @@ func (b *castJSONAsArrayFunctionSig) evalJSON(row chunk.Row) (res types.BinaryJS
...
+			return types.ConvertJSONToInt(sc, item, mysql.HasUnsignedFlag(tp.GetFlag()), tp.GetType())
+		}
+	case types.ETReal, types.ETDecimal:
+		return func(sc *stmtctx.StatementContext, item types.BinaryJSON) (any, error) {
+			if item.TypeCode != types.JSONTypeCodeInt64 && item.TypeCode != types.JSONTypeCodeUint64 && item.TypeCode != types.JSONTypeCodeFloat64 {
+				return nil, ErrInvalidJSONForFuncIndex
+			}
+			return types.ConvertJSONToFloat(sc, item)
+		}
+	case types.ETDatetime:
```
Comment (YangKeao) [inline expression/builtin_cast.go:521]: Could we also support `case mysql.TypeTimestamp`?

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

File: expression/builtin_cast.go (hunk 2 above)
Comment (xiongjiwei) [reply expression/builtin_cast.go:521]: `timestamp` is not able to be the target type of the cast function

**Review by xhebox — APPROVED** [review]

**Review by YangKeao — APPROVED** [review]
LGTM

**Pull request merged**